src/core/namespace.c

```diff
@@ -3310,6 +3312,373 @@ bool ns_type_supported(NamespaceType type) {
         return access(ns_proc, F_OK) == 0;
 }
 
+static int unpeel_get_fd(const char *mount_path, int *ret_fd) {
+        _cleanup_close_pair_ int pipe_fds[2] = EBADF_PAIR;
+        pid_t pid;
+        int r;
+
+        assert(mount_path);
+        assert(ret_fd);
+
+        r = socketpair(AF_UNIX, SOCK_DGRAM|SOCK_CLOEXEC, 0, pipe_fds);
+        if (r < 0)
+                return log_debug_errno(errno, "Failed to create socket pair: %m");
+
+        r = safe_fork("(sd-ns-unpeel)", FORK_DEATHSIG_SIGTERM|FORK_LOG|FORK_WAIT, &pid);
+        if (r < 0)
+                return r;
+        if (r == 0) {
+                _cleanup_close_ int dir_fd = -EBADF;
+
+                pipe_fds[0] = safe_close(pipe_fds[0]);
+
+                /* Clone mount namespace here to unpeel without affecting live process */
+                r = detach_mount_namespace();
```

Comment:
instead of detach_mount_namespace() you can also include FORK_NEW_MOUNTNS and FORK_MOUNTNS_SLAVE in the flags you pass to safe_fork(). It does almost the same (i.e. open a new namespace and remount the while tree to MS_SLAVE; it doesn't remount MS_SHARED afterwards, but we don't need that here)